### PR TITLE
ufal/cannot-upload-big-file

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  0  |  3.5  |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java [hunk 1]
```diff
@@ -53,29 +58,42 @@ public void put(Bitstream bitstream, InputStream in) throws IOException {
...
         //Copy istream to temp file, and send the file, with some metadata
         File scratchFile = File.createTempFile(bitstream.getInternalId(), "s3bs");
-        try {
-            FileUtils.copyInputStreamToFile(in, scratchFile);
-            long contentLength = scratchFile.length();
-            // The ETag may or may not be and MD5 digest of the object data.
-            // Therefore, we precalculate before uploading
-            String localChecksum = org.dspace.curate.Utils.checksum(scratchFile, CSA);
+        try (
+                FileOutputStream fos = new FileOutputStream(scratchFile);
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java:62]: is the in.close() after this important?
it is strange to close InputStream here, or?

File: dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java [hunk 2]
```diff
@@ -53,29 +58,42 @@ public void put(Bitstream bitstream, InputStream in) throws IOException {
...
+                FileOutputStream fosLocalFile = new FileOutputStream(localFile);
+                Utils.bufferedCopy(fisScratchFile, fosLocalFile);
+                fisScratchFile.close();
             }
         } catch (AmazonClientException | IOException | InterruptedException e) {
             log.error("put(" + bitstream.getInternalId() + ", is)", e);
             throw new IOException(e);
+        } catch (NoSuchAlgorithmException nsae) {
+            // Should never happen
+            log.warn("Caught NoSuchAlgorithmException", nsae);
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java:96]: ?

File: dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java [hunk 3]
```diff
@@ -98,4 +116,26 @@ public void remove(Bitstream bitstream) throws IOException {
...
+     * @param localFile
+     * @throws IOException
+     */
+    private void createFileIfNotExist(File localFile) throws IOException {
+        if (localFile.exists()) {
+            return;
+
+        }
+
+        // Create the necessary parent directories
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java:131]: what if parent dir does already exist?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java (hunk 2 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java:96]: It's from Vanilla

File: dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java (hunk 3 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java:131]: It won't be created

File: dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java (hunk 1 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/storage/bitstore/SyncS3BitStoreService.java:62]: It is from Vanilla I just copied it

**Review by vidiecan — APPROVED** [review]

**Pull request merged**